engine/runtime/src/main/java/org/enso/interpreter/node/expression/builtin/meta/IsValueOfTypeNode.java

```diff
@@ -49,6 +49,42 @@ boolean doPolyglotExceptionProxy(Object expectedType, PolyglotExceptionProxy pro
     return isSameObject.execute(expectedType, tpeOfPayload);
   }
 
+  @Specialization(guards = "isAnyType(expectedType)")
+  boolean doAnyType(Object expectedType, Object payload) {
+    return true;
+  }
+
+  @Specialization
+  boolean doLongCheck(Type expectedType, long payload) {
+    var numbers = EnsoContext.get(this).getBuiltins().number();
+    return checkParentTypes(numbers.getSmallInteger(), expectedType);
+  }
+
+  @Specialization
+  boolean doDoubleCheck(Type expectedType, double payload) {
+    var numbers = EnsoContext.get(this).getBuiltins().number();
+    return checkParentTypes(numbers.getDecimal(), expectedType);
+  }
+
+  @Specialization
+  boolean doBigIntegerCheck(Type expectedType, EnsoBigInteger value) {
+    var numbers = EnsoContext.get(this).getBuiltins().number();
+    return checkParentTypes(numbers.getBigInteger(), expectedType);
+  }
+
+  @ExplodeLoop
+  private boolean checkParentTypes(Type actual, Type real) {
+    for (; ; ) {
```

Comment:
could we make it a while loop with `actual` check, please? I'm allergic to (potentially) infinite loops

Reply:
It is not infinite loop as value of `actual` is decreasing with every loop iteration. Moreover it is covered by tests.

Reply:
I see `for(;;)` as a synonym for loop with breaks for languages that don't have (like Rust) [loop](https://doc.rust-lang.org/reference/expressions/loop-expr.html#infinite-loops) construct directly.